src/Orleans/Runtime/OutsideRuntimeClient.cs

```diff
@@ -136,8 +141,25 @@ public OutsideRuntimeClient(ClientConfiguration cfg, GrainFactory grainFactory,
 
             try
             {
-                LoadAdditionalAssemblies();
-                
+                AssemblyCatalog = assemblyCatalog;
+
+                if (AssemblyCatalog == null)
+                {
+                    if (config.Assemblies == null ||
+                        config.Assemblies.Count == 0)
+                        throw new InvalidOperationException("No IAssemblyCatalog implementation or assembly list were provided.");
```

Comment:
> Given that we are now traversing dependencies, it would be nice to also give the option to not specify any assembly, and if that's the case, then it will assume Assembly.GetEntryAssembly() is in the list. So if your host is referencing your grain interfaces project, it would be picked up automatically.

Agreed! Will add that.

> It would be nice to verify if there is any grain implementation in the referenced projects though.
This way for most people this migration is non-breaking, unless they happen to be using OrleansHost.exe, where in that case the host doesn't explicitly reference the implementations

This check will fall push us back to the "Magic Scan" that we have today and not to mention that it is not the Catalog/Loader/Processor responsibility to inspect the assembly content IMHO. This filter to detect if there is a grain, stream, etc, is already done by the codegen IIRC...